internal/client/consensus/grandpa/warp_sync.go

```diff
@@ -0,0 +1,177 @@
+// Copyright 2024 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package grandpa
+
+import (
+	"fmt"
+
+	"github.com/ChainSafe/gossamer/dot/types"
+	"github.com/ChainSafe/gossamer/internal/primitives/core/hash"
+	"github.com/ChainSafe/gossamer/internal/primitives/runtime"
+	"github.com/ChainSafe/gossamer/lib/common"
+	"github.com/ChainSafe/gossamer/pkg/scale"
+)
+
+const MAX_WARP_SYNC_PROOF_SIZE = 8 * 1024 * 1024
+
+var (
+	errMissingStartBlock      = fmt.Errorf("missing start block")
+	errStartBlockNotFinalized = fmt.Errorf("start block is not finalized")
+)
+
+type BlockState interface {
+	GetHeader(hash common.Hash) (*types.Header, error)
+	BestBlockHeader() (*types.Header, error)
+	GetHeaderByNumber(num uint) (*types.Header, error)
+	GetJustification(hash common.Hash) ([]byte, error)
+}
+
+type GrandpaState interface {
+	GetAuthoritesChangesFromBlock(blockNumber uint) ([]uint, error)
+}
+
+type WarpSyncFragment struct {
+	// The last block that the given authority set finalized. This block should contain a digest
+	// signalling an authority set change from which we can fetch the next authority set.
+	Header types.Header
+	// A justification for the header above which proves its finality. In order to validate it the
+	// verifier must be aware of the authorities and set id for which the justification refers to.
+	Justification GrandpaJustification[hash.H256, uint64]
+}
+
+type WarpSyncProof struct {
+	Proofs []WarpSyncFragment
+	// indicates whether the warp sync has been completed
+	IsFinished   bool
+	proofsLength int
+}
+
+func NewWarpSyncProof() *WarpSyncProof {
+	return &WarpSyncProof{
+		Proofs:       make([]WarpSyncFragment, 0),
+		IsFinished:   false,
+		proofsLength: 0,
+	}
+}
+
+func (w *WarpSyncProof) encode() ([]byte, error) {
+	if w == nil {
+		return nil, fmt.Errorf("cannot encode nil WarpSyncProof")
+	}
+	return scale.Marshal(*w)
+}
+
+func (w *WarpSyncProof) addFragment(fragment WarpSyncFragment) (limitReached bool, err error) {
+	encodedFragment, err := scale.Marshal(fragment)
+	if err != nil {
+		return false, err
+	}
+
+	if w.proofsLength+len(encodedFragment) >= MAX_WARP_SYNC_PROOF_SIZE {
+		return true, nil
+	}
+
+	w.proofsLength += len(encodedFragment)
+	w.Proofs = append(w.Proofs, fragment)
+
+	return false, nil
+}
+
+func (w *WarpSyncProof) lastProofBlockNumber() uint64 {
+	return w.Proofs[len(w.Proofs)-1].Justification.Justification.Commit.TargetNumber + 1
+}
+
+type WarpSyncProofProvider struct {
+	blockState   BlockState
+	grandpaState GrandpaState
+}
+
+// Generate build a warp sync encoded proof starting from the given block hash
+func (np *WarpSyncProofProvider) Generate(start common.Hash) ([]byte, error) {
+	// Get and traverse all GRANDPA authorities changes from the given block hash
+	beginBlockHeader, err := np.blockState.GetHeader(start)
+	if err != nil {
+		return nil, errMissingStartBlock
```

Comment:
I don't remember the Go version number that brings this but we can return a chain of errors now, so would be nice to have the error from `GetHeader` appended to the `errMissingStartBlock`
```suggestion
		return nil, fmt.Errorf("%w: %w", errMissingStartBlock, err)
```